File: classic1D/classic_1D.m
function classic_1D

N = 100000;       %总粒子数
Nstep = 200;      %总步数

X = round(linspace(-Nstep,Nstep,2*Nstep+1));  %粒子所有可能到达的位置(横轴)
x = zeros(1,N);  %每个粒子的位置

%> 固定步数Nstep, 观察最后的粒子分布
% tic;
% r = rand(Nstep,N);
% Y = zeros(1,2*Nstep+1);  %统计最终位置上的粒子数
% for i = 1:N
%     for j = 1:Nstep
%         if(r(j,i) < 0.5)
%             x(i) = x(i)-1;
%         else
%             x(i) = x(i)+1;
%         end
%     end
%     index =  find(X == x(i));
%     Y(index) = Y(index) + 1;
% end
% p = Y/N;
% toc

% plot(X,p);


%> 观察粒子分布随时间(步数)的演化
tic;
r = rand(N,Nstep);
Y = zeros(Nstep,2*Nstep+1);  %统计最终位置上的粒子数
for i = 1:Nstep
    for j = 1:N
        if(r(j,i)<0.5)
            x(j) = x(j)-1;
        else
            x(j) = x(j)+1;
        end
        index = find(X==x(j));
        Y(i,index) = Y(i,index) + 1;
    end
end
p = Y/N;
toc


%> 单个粒子随机游走
tic;
x0 = 0;
r0 = rand(1,Nstep);
site = zeros(Nstep,2*Nstep+1);
for i = 1:Nstep
    if(r0(i)<0.5)
        x0 = x0-1;
    else
        x0 = x0+1;
    end
    index = find(X==x0);
    site(i,index) = 1;
end
toc

save("C:\Users\JAH\Desktop\组\random walk\classic1D\classic1D.mat","p","X","Nstep","site");
fprintf("数据已导出");

File: classic1D/classic_1D.py
from os import X_OK
import numpy as np
import matplotlib.pyplot as plt
import matplotlib as mpl
import scipy.io as sio
from matplotlib import animation
from matplotlib.animation import FuncAnimation
from matplotlib import colors

data = sio.loadmat(r'C:\Users\JAH\Desktop\组\random walk\classic1D\classic1D.mat')

p = np.array(data["p"])
X = np.array(data["X"])[0]
site = np.array(data["site"])
Nstep = int(data["Nstep"])

fig, ax1 = plt.subplots()


#> 观察粒子分布随时间(步数)的演化

# def update(i):

#     ax1.clear()    #清除上一帧的axes对象
#     ax1.plot(X,p[i],linewidth=0.2)

#     ax1.set_title("1D classic random walk")
#     ax1.set_xlabel('X')
#     ax1.set_ylabel('p')

#     # ax1.set_xlim(-Nstep,Nstep)
#     # ax1.set_ylim(0,1)

#     return ax1


#> 单个粒子随机游走

X0, Y0 = np.meshgrid(X,X)
cMap = colors.ListedColormap(['white','black']) #自定义颜色 黑白色
norm = mpl.colors.Normalize(0,1)

def update(i):

    ax1.clear()    #清除上一帧的axes对象

    Z0 = np.zeros((2*Nstep+1,2*Nstep+1))
    Z0[Nstep] = site[i]
    g = ax1.pcolor(X0, Y0, Z0, norm=norm, cmap=cMap)

    ax1.set_title("1D classic random walk")
    ax1.set_xlabel('X')

    ax1.set_xlim(-Nstep/5,Nstep/5)
    ax1.set_ylim(-Nstep/5,Nstep/5)

    return ax1


anim = FuncAnimation(fig, update, frames=np.arange(0, Nstep), interval=50, blit=False, repeat=False)
# plt.show()
anim.save(r'C:\\Users\\JAH\\Desktop\\组\\random walk\\classic1D\\1D classic single particle.gif')